### DISCUSSION TITLE
synthesis of the output speech in the finetinung loop

### DISCUSSION THREAD

**Discussion by laurentbenaroya** [QwenLM/Qwen3-TTS#271] [General]
Hello,
Is it possible to synthesize speech audio in the training loop for fine-tuning, from the logits (outputs.logits and sub_talker_logits) ?
To provide some context, here is  part of the code in `finetuning/sft_12hz.py`:

```
outputs = model.talker(
    inputs_embeds=input_embeddings[:, :-1, :],
    attention_mask=attention_mask[:, :-1],
    labels=codec_0_labels[:, 1:],
    output_hidden_states=True
)

hidden_states = outputs.hidden_states[0][-1]
talker_hidden_states = hidden_states[codec_mask[:, 1:]]
talker_codec_ids = codec_ids[codec_mask]

sub_talker_logits, sub_talker_loss = model.talker.forward_sub_talker_finetune(talker_codec_ids, talker_hidden_states)

loss = outputs.loss + 0.3 * sub_talker_loss
```
I would like to reconstruct the speech signal from the model, perhaps using the Qwen3TTSTokenizer class or a lower-level approach using VQ and RVQ.
Thanks in advance for your help,
